Restrict the set of defaults-overridable flags to better match shipping Safari

brentfulgham opened the pull request:
#### 0f58ae8a9cfeded106727881a403f817a692d8f4
<pre>
Restrict the set of defaults-overridable flags to better match shipping Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=251183">https://bugs.webkit.org/show_bug.cgi?id=251183</a>
&lt;rdar://104657361&gt;

Reviewed by Elliott Williams.

The new feature generator logic attempts to persist all `internal` feature flags such
that a developer can use a &apos;WebKitDebug&apos;-prefix for a setting to override it. This is
only desired for a small set of feature flags (those that lived in the
Source/WTF/Scripts/Preferences/WebPreferencesDebug.yaml file previously).

This patch does the following:

1. Corrects a handful of feature flags that were moved from `internal` to `embedder`, improperly.
2. Adds a new key, `defaultsOverridable`, to UnifiedWebPreferences.yaml. Flags that
have this key with a `true` value will get additional logic to support
&apos;WebKitDebug&apos;-prefix overrides.
3. Changes the predicate from &apos;defaultOverridable?&apos; to `defaultsOverridable?&apos; for consistency.
4. Partial revert of 258448@main to restore the ability to honor debug overrides for UInt32 types.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb:
* Source/WebKit/UIProcess/Cocoa/WebPreferencesCocoa.mm:
(WebKit::setDebugUInt32ValueIfInUserDefaults): Partial revert of 258448@main. We actually
do need this method.

Canonical link: <a href="https://commits.webkit.org/259458@main">https://commits.webkit.org/259458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ce2257391425e5dd61b8438f718e83b552931af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104930 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14010 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37827 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114196 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174389 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15143 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4935 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97253 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113217 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110690 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108391 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26329 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94787 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7355 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27688 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92818 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5085 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7447 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30247 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13504 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47240 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101512 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6526 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9238 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25321 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WTF/Scripts/GeneratePreferences.rb:
```diff
@@ -188,11 +190,11 @@ def upcast
+    %w{ unstable testable }.include? @status
   end
 
   def defaultOverridable?
```
While you're here, would you mind renaming this to `defaultsOverridable?`, too? It should be only used in one template.

brentfulgham replied:
Sure! And it looks like I need to add the casting stuff we used elsewhere. I’ll upload a new patch in the morning.

brentfulgham replied:
Actually, I actually just need to restore the Uint32 version of the setting method that was removed in [258448@main](https://commits.webkit.org/258448@main)

patrickangle reviewed (commented):
lgtm, but I trust Elliott more than me on this

emw-apple reviewed (approved):
Let's add a note to the documentation at the top of UWP.yaml. Maybe something like:

```diff
diff --git a/Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml b/Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml
index 1d863efb6b4c..00ee00ef3032 100644
--- a/Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml
+++ b/Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml
@@ -52,7 +52,9 @@
 #   OFF by default.
 # internal::
 #   Tools for debugging the WebKit engine. Not generally useful to web developers.
-#   OFF by default.
+#   OFF by default. When declared with "defaultsOverridable: true", platform
+#   user defaults may override a default value at runtime, using the preference
+#   name prefixed with "WebKitDebug".
 #
 # == Embedder Controls ==
 # embedder::
```

emw-apple commented on Source/WebKit/UIProcess/Cocoa/WebPreferencesCocoa.mm:
```diff
@@ -130,6 +130,17 @@ static void setDebugBoolValueIfInUserDefaults(const String& identifier, const St
     store.setBoolValueForKey(key, [object boolValue]);
 }
 
+static void setDebugUInt32ValueIfInUserDefaults(const String& identifier, const String& keyPrefix, const String& globalDebugKeyPrefix, const String& key, WebPreferencesStore& store)
+{
+    id object = debugUserDefaultsValue(identifier, keyPrefix, globalDebugKeyPrefix, key);
+    if (!object)
+        return;
+    if (![object respondsToSelector:@selector(unsignedIntegerValue)])
+        return;
+
+    store.setUInt32ValueForKey(key, [object unsignedIntegerValue]);
+}
+
```
Thanks for bringing this back, confirmed the implementation matches what I removed in 9def6e6f0258ddd9171d32bec2fffb684dbc7b8b

brentfulgham committed:
Restrict the set of defaults-overridable flags to better match shipping Safari
https://bugs.webkit.org/show_bug.cgi?id=251183
<rdar://104657361>

Reviewed by Elliott Williams.

The new feature generator logic attempts to persist all `internal` feature flags such
that a developer can use a 'WebKitDebug'-prefix for a setting to override it. This is
only desired for a small set of feature flags (those that lived in the
Source/WTF/Scripts/Preferences/WebPreferencesDebug.yaml file previously).

This patch does the following:

1. Corrects a handful of feature flags that were moved from `internal` to `embedder`, improperly.
2. Adds a new key, `defaultsOverridable`, to UnifiedWebPreferences.yaml. Flags that
have this key with a `true` value will get additional logic to support
'WebKitDebug'-prefix overrides.
3. Changes the predicate from 'defaultOverridable?' to `defaultsOverridable?' for consistency.
4. Partial revert of 258448@main to restore the ability to honor debug overrides for UInt32 types.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb:
* Source/WebKit/UIProcess/Cocoa/WebPreferencesCocoa.mm:
(WebKit::setDebugUInt32ValueIfInUserDefaults): Partial revert of 258448@main. We actually
do need this method.

Canonical link: https://commits.webkit.org/259458@main
